tests/bsim/bluetooth/samples/battery_service/src/central_test.c

```diff
@@ -0,0 +1,338 @@
+
+static struct bt_gatt_subscribe_params battery_level_notify_params;
+static struct bt_gatt_subscribe_params battery_level_status_sub_params;
+
+/*
+ * Battery Service  test:
+ *   We expect to find a connectable peripheral to which we will
+ *   connect and discover Battery Service
+ *
+ *   Test the Read/Notify/Indicate Characteristics of BAS
+ */
+
+#define WAIT_TIME                  10 /*seconds*/
+#define BAS_BLS_IND_RECEIVED_COUNT 20
+#define BAS_BLS_NTF_RECEIVED_COUNT 20
+
+static DEFINE_FLAG(notification_count_reached);
+static DEFINE_FLAG(indication_count_reached);
+
+extern enum bst_result_t bst_result;
+
+static void test_bas_central_init(void)
+{
+	bst_ticker_set_next_tick_absolute(WAIT_TIME * 1e6);
+	bst_result = In_progress;
+}
+
+static void test_bas_central_tick(bs_time_t HW_device_time)
+{
+	/*
+	 * If in WAIT_TIME seconds the testcase did not already pass
+	 * (and finish) we consider it failed
+	 */
+	if (bst_result != Passed) {
+		TEST_FAIL("test_bas_central failed (not passed after %i seconds)\n", WAIT_TIME);
+	}
+}
+
+/* Callback for handling Battery Level Notifications */
+static uint8_t battery_level_notify_cb(struct bt_conn *conn,
+				       struct bt_gatt_subscribe_params *params, const void *data,
+				       uint16_t length)
+{
+	if (data) {
+		LOG_INF("[NOTIFICATION] BAS Battery Level: %d%%", *(const uint8_t *)data);
+	} else {
+		LOG_INF("Battery Level Notifications disabled");
+	}
+	return BT_GATT_ITER_CONTINUE;
+}
+
+/* Callback for handling Battery Level Read Response */
+static uint8_t battery_level_read_cb(struct bt_conn *conn, uint8_t err,
+				     struct bt_gatt_read_params *params, const void *data,
+				     uint16_t length)
+{
+	TEST_ASSERT(err == 0, "Failed to read Battery Level (err %d)", err);
+	if (data) {
+		LOG_DBG("[READ] BAS Battery Level: %d%%\n", *(const uint8_t *)data);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static unsigned char battery_level_status_indicate_cb(struct bt_conn *conn,
+						      struct bt_gatt_subscribe_params *params,
+						      const void *data, uint16_t length)
+{
+	static int ind_received;
+
+	if (!data) {
+		LOG_INF("bas level status indication disabled\n");
+	} else {
+		printk("[INDICATION]  BAS Battery Level Status: ");
+		for (int i = 0; i < length; i++) {
+			printk("%02x ", ((uint8_t *)data)[i]);
+		}
+		printk("\n");
+
+		if (ind_received++ > BAS_BLS_IND_RECEIVED_COUNT) {
+			SET_FLAG(indication_count_reached);
+		}
+	}
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t battery_level_status_notify_cb(struct bt_conn *conn,
+					      struct bt_gatt_subscribe_params *params,
+					      const void *data, uint16_t length)
+{
+	static int notify_count;
+
+	if (!data) {
+		LOG_INF("[UNSUBSCRIBED]\n");
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
+
+	printk("[NOTIFICATION] BAS Battery Level Status: ");
+	for (int i = 0; i < length; i++) {
+		printk("%02x ", ((uint8_t *)data)[i]);
+	}
+	printk("\n");
```

Comment:
I have an idea that the implementation isn't correct due to the use of `WRITE_BIT`. 

We should actually verify the content in the test rather than just accepting _any_ notification.

Reply:
Ditto for the indication (both can/should use the same parser)